File: src/main/java/org/minefortress/tasks/ServerTaskManager.java
package org.minefortress.tasks;


import net.minecraft.entity.Entity;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IWorkerPawn;
import net.remmintan.mods.minefortress.core.interfaces.server.ITickableManager;
import net.remmintan.mods.minefortress.core.interfaces.server.IWritableManager;
import net.remmintan.mods.minefortress.core.interfaces.tasks.IInstantTask;
import net.remmintan.mods.minefortress.core.interfaces.tasks.IServerTaskManager;
import net.remmintan.mods.minefortress.core.interfaces.tasks.ITask;
import net.remmintan.mods.minefortress.core.utils.CoreModUtils;
import net.remmintan.mods.minefortress.networking.helpers.FortressServerNetworkHelper;
import net.remmintan.mods.minefortress.networking.s2c.S2CAddClientTasksPacket;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public class ServerTaskManager implements IServerTaskManager, IWritableManager, ITickableManager {
    private final Map<UUID, ITask> nonFinishedTasks = new HashMap<>();
    private final Queue<ITask> notStartedTasks = new LinkedList<>();

    private static @NotNull List<IWorkerPawn> filterWorkers(List<Integer> selectedPawnIds, ServerPlayerEntity player) {
        final var serverWorld = player.getWorld();
        return selectedPawnIds
                .stream()
                .map(serverWorld::getEntityById)
                .filter(IWorkerPawn.class::isInstance)
                .map(IWorkerPawn.class::cast)
                .filter(worker -> !worker.getTaskControl().isDoingEverydayTasks())
                .toList();
    }

    @Override
    public void addTask(ITask task, List<Integer> selectedPawnIds, ServerPlayerEntity player) {
        removeAllFinishedTasks();

        final var packet = new S2CAddClientTasksPacket(task.toTaskInformationDto());
        FortressServerNetworkHelper.send(player, S2CAddClientTasksPacket.CHANNEL, packet);
        nonFinishedTasks.put(task.getId(), task);

        if (selectedPawnIds.isEmpty()) {
            notStartedTasks.add(task);
            return;
        }

        final var selectedWorkers = filterWorkers(selectedPawnIds, player);
        if (selectedWorkers.isEmpty()) {
            player.sendMessage(Text.of("No appropriate workers selected. Put the task in the queue"), false);
            notStartedTasks.add(task);
            return;
        }

        task.prepareTask();
        setPawnsToTask(task, selectedWorkers);
    }

    @Override
    public void tick(@Nullable ServerPlayerEntity player) {
        if (player == null) return;
        if (notStartedTasks.isEmpty()) return;
        final var freeWorkers = CoreModUtils.getFortressManager(player).getFreeWorkers();
        if (freeWorkers.size() > 2) {
            final var task = notStartedTasks.remove();
            final var freeWorkersIds = freeWorkers.stream().map(it -> ((Entity) it).getId()).toList();
            this.addTask(task, freeWorkersIds, player);
        }
    }

    @Override
    public void executeInstantTask(IInstantTask task, ServerPlayerEntity player) {
        task.execute(player.getServerWorld(), player, getManagersProvider(player)::getBuildingsManager);
    }

    @Override
    public void cancelTask(UUID id, ServerPlayerEntity player) {
        removeAllFinishedTasks();
        final var removedTask = nonFinishedTasks.remove(id);
        if(removedTask != null)
            removedTask.cancel();
        getManagersProvider(player).getResourceManager().returnReservedItems(id);
    }

    private void removeAllFinishedTasks() {
        final var finishedTasks = nonFinishedTasks.entrySet()
                .stream()
                .filter(e -> e.getValue().taskFullyFinished())
                .toList();
        finishedTasks.forEach(e -> nonFinishedTasks.remove(e.getKey()));
    }

    private void setPawnsToTask(ITask task, List<IWorkerPawn> workers) {
        for(IWorkerPawn worker : workers) {
            if(!task.hasAvailableParts()) break;
            worker.getTaskControl().setTask(task);
        }
    }


    @Override
    public void write(NbtCompound tag) {

    }

    @Override
    public void read(NbtCompound tag) {

    }
}

File: src/main/java/org/minefortress/tasks/ServerTaskManager.kt
package org.minefortress.tasks

import net.minecraft.entity.Entity
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NbtCompound
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.remmintan.mods.minefortress.blocks.FortressBlocks
import net.remmintan.mods.minefortress.blocks.task.FortressTaskBlockEntity
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IWorkerPawn
import net.remmintan.mods.minefortress.core.interfaces.server.IServerManagersProvider
import net.remmintan.mods.minefortress.core.interfaces.server.ITickableManager
import net.remmintan.mods.minefortress.core.interfaces.server.IWritableManager
import net.remmintan.mods.minefortress.core.interfaces.tasks.*
import net.remmintan.mods.minefortress.core.utils.LogCompanion
import net.remmintan.mods.minefortress.core.utils.ServerModUtils
import net.remmintan.mods.minefortress.core.utils.getManagersProvider
import net.remmintan.mods.minefortress.core.utils.isSurvivalFortress
import net.remmintan.mods.minefortress.networking.helpers.FortressChannelNames
import net.remmintan.mods.minefortress.networking.helpers.FortressServerNetworkHelper
import net.remmintan.mods.minefortress.networking.s2c.ClientboundTaskExecutedPacket
import net.remmintan.mods.minefortress.networking.s2c.S2CAddClientTasksPacket
import java.util.*
import java.util.concurrent.ConcurrentHashMap


class ServerTaskManager(private val server: MinecraftServer, fortressPos: BlockPos) : IServerTaskManager,
    IWritableManager,
    ITickableManager {
    private val tasksInProgress: MutableMap<BlockPos, IBaseTask> = ConcurrentHashMap()
    private val notStartedTasks: Queue<IBaseTask> = LinkedList()

    private val managersProvider: IServerManagersProvider by lazy { server.getManagersProvider(fortressPos) }
    private val world: ServerWorld by lazy { server.overworld }

    override fun addTask(task: IBaseTask, selectedPawnIds: List<Int>, player: ServerPlayerEntity) {
        placeTaskInTheWorldAndReserveItems(task, player)

        //TODO place the task in the world + reserve items

        if (selectedPawnIds.isEmpty()) {
            notStartedTasks.add(task)
            return
        }

        val selectedWorkers = filterWorkers(selectedPawnIds, player)
        if (selectedWorkers.isEmpty()) {
            notStartedTasks.add(task)
            return
        }

        if (task is ITaskWithPreparation)
            task.prepareTask()
        tasksInProgress[task.pos] = task
        setPawnsToTask(task, selectedWorkers)
    }

    override fun tick(server: MinecraftServer, world: ServerWorld, player: ServerPlayerEntity?) {
        removeAllFinishedTasks()

        //FIXME take planned tasks even when the player is offline
        if (player == null) return
        if (notStartedTasks.isEmpty() && tasksInProgress.isEmpty()) return
        val readyWorkers = ServerModUtils
            .getFortressManager(player)
            .map { it.readyWorkers }
            .orElse(emptyList())

        if (readyWorkers.isNotEmpty()) {
            val inProgressTask = tasksInProgress.values
                .firstOrNull { it.notCancelled() && it.canTakeMoreWorkers() }
                ?.also { setPawnsToTask(it, readyWorkers) }

            if (inProgressTask == null) {
                notStartedTasks.poll()?.let {
                    val selectedPawnIds = readyWorkers.map { w -> (w as Entity).id }
                    this.addTask(it, selectedPawnIds, player)
                }
            }
        }
    }

    override fun cancelTask(pos: BlockPos, player: ServerPlayerEntity) {
        val removedTask = tasksInProgress.remove(pos)
        removedTask?.cancel()
        val resourceManager = managersProvider.resourceManager
        val resourceHelper = managersProvider.resourceHelper
        if (!resourceHelper.transferItemsFromTask(resourceManager, pos)) {
            log.error("Couldn't return items from task at $pos to the resource manager")
        }
        world.removeBlock(pos, false)
    }

    private fun removeAllFinishedTasks() {
        val finishedTasks = tasksInProgress.filterValues { it.isComplete() }.keys
        finishedTasks.forEach {
            tasksInProgress.remove(it)
            world.removeBlock(it, false)
        }
    }

    private fun setPawnsToTask(task: IBaseTask, workers: List<IWorkerPawn>) {
        when (task) {
            is ITask ->
                for (worker in workers) {
                    if (!task.hasAvailableParts() || !task.canTakeMoreWorkers()) break
                    task.addWorker()
                    worker.taskControl.setTask(task)
                }

            is IAreaBasedTask ->
                for (worker in workers) {
                    if (!task.hasMoreBlocks() || !task.canTakeMoreWorkers()) break
                    task.addWorker()
                    worker.areaBasedTaskControl.setTask(task)
                }

            else -> error("Wrong task class")
        }
    }


    override fun write(tag: NbtCompound) {
    }

    override fun read(tag: NbtCompound) {
    }

    private fun placeTaskInTheWorldAndReserveItems(task: IBaseTask, player: ServerPlayerEntity) {
        val taskInformationDto = task.toTaskInformationDto()

        val taskPos = task.pos
        if (world.getBlockEntity(taskPos) !is FortressTaskBlockEntity) {
            world.setBlockState(taskPos, FortressBlocks.FORTRESS_TASK.defaultState)
            val success = reserveItems(task, taskPos)
            if (!success) {
                FortressServerNetworkHelper.send(
                    player,
                    FortressChannelNames.FINISH_TASK,
                    ClientboundTaskExecutedPacket(taskPos)
                )
            }
        }

        val packet = S2CAddClientTasksPacket(taskInformationDto)
        FortressServerNetworkHelper.send(player, S2CAddClientTasksPacket.CHANNEL, packet)
    }

    private fun reserveItems(task: IBaseTask, taskPos: BlockPos): Boolean {
        if (!server.isSurvivalFortress()) return true

        val resourceHelper = managersProvider.resourceHelper
        val resourceManager = managersProvider.resourceManager
        val success = when (task) {
            is SimpleSelectionTask -> {
                task.placingItem?.let {
                    val item = it
                    val amount = task.positions.size
                    val stack = ItemStack(item, amount)
                    resourceHelper.transferItemsToTask(
                        resourceManager,
                        taskPos,
                        listOf(stack)
                    )
                } ?: true
            }

            is AreaBlueprintTask -> {
                resourceHelper.transferItemsToTask(
                    resourceManager,
                    taskPos,
                    task.requiredResources
                )
            }

            is RepairBuildingTask -> {
                resourceHelper.transferItemsToTask(
                    resourceManager,
                    taskPos,
                    task.repairItems
                )
            }

            else -> true
        }

        if (!success) {
            log.error("Was not able to reserve items for task $taskPos. Seemingly not enough items")
        }

        return success
    }

    companion object : LogCompanion(ServerTaskManager::class) {
        private fun filterWorkers(selectedPawnIds: List<Int>, player: ServerPlayerEntity): List<IWorkerPawn> {
            val serverWorld = player.world
            return selectedPawnIds
                .asSequence()
                .map { serverWorld.getEntityById(it) }
                .filterNotNull()
                .filter { it is IWorkerPawn }
                .map { it as IWorkerPawn }
                .filter { !it.taskControl.isDoingEverydayTasks }
                .toList()
        }
    }
}
